lib/modules/manager/pipenv/artifacts.ts

```diff
@@ -113,35 +114,95 @@ export function getPipenvConstraint(
+  const parsedUrl = parseUrl(url);
+  if (parsedUrl) {
+    parsedUrl.username = '';
+    parsedUrl.password = '';
+    const urlWithoutCredentials = parsedUrl.toString();
+
+    return find({ hostType: PypiDatasource.id, url: urlWithoutCredentials });
+  }
 }
 
-async function findPipfileSourceUrlWithCredentials(
+async function findPipfileSourceUrlsWithCredentials(
   pipfileContent: string,
   pipfileName: string,
-): Promise<string | null> {
+): Promise<string[]> {
   const pipfile = await extractPackageFile(pipfileContent, pipfileName);
-  if (!pipfile) {
-    logger.debug('Error parsing Pipfile');
+
+  return (
+    pipfile?.registryUrls?.filter((url) => !!parseUrl(url)?.username) ?? []
+  );
+}
+
+/**
+ * This will extract the actual variable name from an environment-placeholder:
+ * ${USERNAME:-defaultvalue} will yield 'USERNAME'
+ */
+export function extractEnvironmentVariableName(
+  credential: string | null,
+): string | null {
+  if (!credential) {
     return null;
   }
+  const match = decodeURI(credential).match('([a-zA-Z0-9_]+)');
+  return match?.length ? match[0] : null;
+}
 
-  const credentialTokens = [
-    '$USERNAME:',
-    // eslint-disable-next-line no-template-curly-in-string
-    '${USERNAME}',
-    '$PASSWORD@',
-    // eslint-disable-next-line no-template-curly-in-string
-    '${PASSWORD}',
-  ];
-
-  const sourceWithCredentials = pipfile.registryUrls?.find((url) =>
-    credentialTokens.some((token) => url.includes(token)),
+export function addExtraEnvVariable(
+  extraEnv: ExtraEnv<unknown>,
+  environmentVariableName: string,
+  environmentValue: string,
+): void {
+  logger.debug(
+    `Adding ${environmentVariableName} environment variable for pipenv`,
   );
+  if (
+    extraEnv[environmentVariableName] &&
+    extraEnv[environmentVariableName] !== environmentValue
+  ) {
+    logger.warn(
+      `Possible misconfiguration, ${environmentVariableName} is already set to a different value`,
+    );
+  }
+  extraEnv[environmentVariableName] = environmentValue;
+}
 
-  // Only one source is currently supported
-  return sourceWithCredentials ?? null;
+/**
+ * Pipenv allows configuring source-urls for remote repositories with placeholders for credentials, i.e. http://$USER:$PASS@myprivate.repo
+ * if a matching host rule exists for that repository, we need to set the corresponding variables.
+ * Simply substituting them in the URL is not an option as it would impact the hash for the resulting Pipfile.lock
+ *
+ */
+async function addCredentialsForSourceUrls(
+  newPipfileContent: string,
+  pipfileName: string,
+  extraEnv: ExtraEnv<unknown>,
+): Promise<void> {
+  const sourceUrls = await findPipfileSourceUrlsWithCredentials(
+    newPipfileContent,
+    pipfileName,
+  );
+  for (const sourceUrl of sourceUrls) {
+    const parsedSourceUrl = parseUrl(sourceUrl);
+    if (!parsedSourceUrl) {
+      logger.warn(`Pipenv: could not parse url for source: ${sourceUrl}`);
+      return;
+    }
```

Comment:
In theory this should never be reached, because the list of sourceUrls was already filtered using a parseUrl?

Reply:
correct. tsc just did not understand. that branch will be gone after next commit.